walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/MeetingScheduler.java

```diff
@@ -0,0 +1,107 @@
+package com.google.sps;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Deque;
+
+/**
+ * Schedules available meeting times based on a meeting request
+ * and a collection of events which have already been scheduled.
+ */
+public class MeetingScheduler {
+  private final Collection<Event> events;
+  private final MeetingRequest request;
+  private ArrayList<TimeRange> availableMeetingTimes;
+  private Event[] eventsOrderedByStartTime;
+  private Deque<TimeRange> busyTimeRanges;
+  private boolean isCached;
+  private int currentTime;
+
+  private MeetingScheduler(Collection<Event> events, MeetingRequest request) {
+    this.events = events;
+    this.request = request;
+    this.availableMeetingTimes = new ArrayList<>();
+    this.eventsOrderedByStartTime = new Event[events.size()];
+    this.busyTimeRanges = new ArrayDeque<>();
+    this.isCached = false;
+    this.currentTime = TimeRange.START_OF_DAY;
+  }
+
+  /**
+   * Returns a meeting scheduler containing the specified events and meeting request.
+   * @param events The specified events.
+   * @param request The specified request.
+   * @return A meeting scheduler containing the specified events and meeting request.
+   * @throws IllegalArgumentException If events or request is null.
+   */
+  public static MeetingScheduler with(Collection<Event> events, MeetingRequest request)
+      throws IllegalArgumentException {
+    if (events == null || request == null) {
+      throw new IllegalArgumentException("events and request must both be non-null.");
+    }
+    return new MeetingScheduler(events, request);
+  }
+
+  /**
+   * Generates and returns a collection of available meeting times.
+   * @return A collection of available meeting times.
+   */
+  public Collection<TimeRange> getAvailableMeetingTimes() {
+    if (!isCached) {
+      orderEventsByStartTime();
+      generateBusyTimeRanges();
+      generateAvailableMeetingTimes();
+      isCached = true;
+    }
+    return availableMeetingTimes;
+  }
+
+  private void orderEventsByStartTime() {
+    eventsOrderedByStartTime = events.toArray(eventsOrderedByStartTime);
+    Arrays.sort(eventsOrderedByStartTime,
+        (event1, event2) -> TimeRange.ORDER_BY_START.compare(event1.getWhen(), event2.getWhen()));
+  }
+
+  private void generateBusyTimeRanges() {
```

Comment:
like this function, this function can not be called multiple times. which defeats its purpose as a function

Reply:
Removed this method and moved the code into the calling method.